active_storage/server/app.py

```diff
@@ -1,162 +1,215 @@
+
+        validate_request(request_data) #Will return a relevant HTTP response to the client if request is invalid
+
+        #Can't use generator funcs with Depends(...) so call here instead
+        upstream_response = upstream_s3_response_generator(request_data, credentials)
+
+        #Process upstream response in chunks
         result = None
-        async for chunk in upstream_response.aiter_bytes():
-            if not chunk:
-                continue
-            nparr = np.frombuffer(chunk, dtype)
-            if result is not None:
-                nparr = np.append(nparr, result)
-            result = reducer(nparr)
-        return Response(result.tobytes(), media_type = "application/octet-stream")
-    return handler
+        async for chunk in upstream_response:
+            np_arr = np.frombuffer(chunk, dtype=request_data.dtype)
+            chunk_result = operation.chunk_reducer(np_arr)
+            if result is None:
+                result = chunk_result #Only happens for first chunk to be processed
+            else:
+                result = operation.aggregator(chunk_result, result)
 
+        response_headers = {
+            'x-activestorage-dtype': str(result.dtype),
+            'x-activestorage-shape': str(list(result.shape)),
+        }
 
-REDUCERS = {
-    "max": np.amax,
-    "min": np.amin,
-    # Prevent sum from using a dtype with increased precision
-    "sum": lambda arr: np.sum(arr, dtype = arr.dtype),
-    # Always return counts as int64
-    "count": lambda arr: np.prod(arr.shape, dtype = np.int64),
-}
-for name, reducer in REDUCERS.items():
-    app.add_api_route(
-        f"/{name}/{{dtype}}/{{obj_path:path}}",
-        make_reducer_handler(reducer),
-        methods = ["GET"]
-    )
+        return Response(content=result.tobytes(), status_code=200, media_type='application/octet-stream', headers=response_headers)
 
+    return handler
 
-@app.get("/obj/{obj_path:path}")
-async def obj(upstream_response: httpx.Response = Depends(upstream_response)):
-    """
-    Returns the content of the object at the given path.
-    """
-    return StreamingResponse(
-        upstream_response.aiter_bytes(),
-        status_code = upstream_response.status_code
-    )
 
 
-@app.get("/.well-known/s3-active-storage")
-async def well_known():
-    """
-    Responds with information about the supported operations.
-    """
-    return {
-        "active_storage_version": "v1",
-        "s3_endpoint": settings.s3_endpoint,
-        "available_reducers": list(REDUCERS.keys()),
-        "supported_datatypes": list(NumericDataType.DATATYPES.keys()),
-    }
+class OctetStreamResponse(Response):
+    """ Dummy response class which ensures that OpenAPI generated schema have correct response type for handler functions """
+    media_type = 'application/octet-stream'
+
+
+#Loop through allowed operations and generate api endpoints
+for op in ALLOWED_OPERATIONS.values():
```

Comment:
I don't think we need this anymore. A single handler with the reducer name as a path parameter would be better.